### ISSUE TITLE
Run migrations in-process without DI

### ISSUE THREAD

**Issue by taprm** [fluentmigrator/fluentmigrator#1208]
Hello!
Is there any chance to create MigrationRunner directly without using DI to apply migrations?
Thanks

**Comment by jzabroski** [comment]
I'm not sure exactly what problem you're trying to solve?

**Comment by taprm** [comment]
> I'm not sure exactly what problem you're trying to solve?

I have an asp.net core application, with autofac as a DI. So I want my application to apply all migrations during start. How can I create necessary instances of migration runners in my case? I thought that there must be a way to use migration runners without dependency injection at all

**Comment by jzabroski** [comment]
I think the path forward here is to pull the example out of the examples folder and turn it into an FluentMigratorApplication abstract base class.  The Application class would hook into whatever DI you inject.  In this way, we could simplify building in-process toolchains which is the current recommended guidance, anyway, given #1016 

Your code would then look like:

```c#
public class AspNetFluentMigratorApp : FluentMigratorApplication
{
  public AspNetFluentMigratorApp(IMigrationRunner migrationRunner) : (migrationRunner) {
    this.Start();
  }
}
```

@fubar-coder What do you think?  It's just pseudo-code at this point and missing the hard battle of getting it working.

**Comment by jzabroski** [comment]
Related (old) question on StackOverflow: https://stackoverflow.com/questions/7574417/is-it-possible-to-use-fluent-migrator-in-application-start

**Comment by jzabroski** [comment]
Technically, some of this infrastructure already pretty much exists, viz a viz:

https://github.com/fluentmigrator/fluentmigrator/blob/782e12cda168c9b94ba5647e3edf713d7a738e03/src/FluentMigrator.DotNet.Cli/Setup.cs#L37-L43

What would need to be cleaned up is the bits where this logic calls BuildServiceProvider, as the [ASP.NET Runtime Host needs to be in charge of doing that correctly](https://twitter.com/davidfowl/status/1038521947007594498), or  the ServiceCollection will get transformed into a ServiceProvider multiple times.  When that happens, each ServiceProvider can have a reference to a different universe of types and some of the "services" may be disposed of by the time FluentMigrator (or the ASP.NET web page) get a chance to use those services.

In other words, the following line needs to change from returning an IServiceProvider to an IServiceCollection, so that the caller can decide whether they need to kick the job of calling `BuildServiceProvider()` to a runtime or handle it themselves:

https://github.com/fluentmigrator/fluentmigrator/blob/782e12cda168c9b94ba5647e3edf713d7a738e03/src/FluentMigrator.DotNet.Cli/Setup.cs#L153

The key idea is this `Setup` class logic is what powers the `BaseCommand` that powers the `dotnet-fm` out-of-process runner:

https://github.com/fluentmigrator/fluentmigrator/blob/782e12cda168c9b94ba5647e3edf713d7a738e03/src/FluentMigrator.DotNet.Cli/Commands/BaseCommand.cs#L25-L34
